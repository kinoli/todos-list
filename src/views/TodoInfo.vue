<template>
  <section class="todo-info">
    <v-layout row>
      <v-flex xs12 sm6 offset-sm3>
        <v-card>
          <v-toolbar dark tabs flat color="teal">
            <v-btn id="btn-back" icon @click="onBack">
              <v-icon>chevron_left</v-icon>
            </v-btn>
            <v-toolbar-title>{{ details.title }}</v-toolbar-title>
            <v-spacer></v-spacer>
          </v-toolbar>
          <v-tabs
            v-if="details && details.user"
            @change="onNewTab"
            v-model="active"
            color="teal"
            dark
            slider-color="yellow"
          >
            <v-tab
              v-for="tab in tabs"
              :key="'tab='+tab"
              ripple
              class="tab"
            >
              {{ tab }}
            </v-tab>
            <v-tab-item :key="tabs[0]">
              <v-card flat>
                <v-card-text>
                  <v-subheader>Todo info</v-subheader>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>ID</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.id }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Title</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.title }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Completed</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.completed }}</v-list-tile-action>
                  </v-list-tile>
                </v-card-text>
              </v-card>
            </v-tab-item>

            <v-tab-item :key="tabs[1]">
              <v-card flat>
                <v-card-text>
                  <v-subheader>User info</v-subheader>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>User ID</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.id }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Name</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.name }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Username</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.username }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Email</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.email }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Phone</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.phone }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Website</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.website }}</v-list-tile-action>
                  </v-list-tile>
                </v-card-text>
              </v-card>
            </v-tab-item>

            <v-tab-item :key="tabs[2]">
              <v-card flat>
                <v-card-text>
                  <v-subheader>Address</v-subheader>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Street</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.address.street }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>City</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.address.city }}</v-list-tile-action>
                  </v-list-tile>
                  <v-list-tile>
                    <v-list-tile-content>
                      <v-list-tile-title>Zip code</v-list-tile-title>
                    </v-list-tile-content>
                    <v-list-tile-action>{{ details.user.address.zipcode }}</v-list-tile-action>
                  </v-list-tile>
                </v-card-text>
              </v-card>
            </v-tab-item>
          </v-tabs>
        </v-card>
      </v-flex>
    </v-layout>
  </section>
</template>

<script lang="js">
export default {
  name: 'todo-info',
  mounted () {
    this.details = this.$store.getters.getTodoDetails(this.$route.params.id)
    this.active = this.$store.getters.getActiveTab
  },
  data: () => ({
    details: {},
    tabs: [
      'Todo',
      'User',
      'Address'
    ],
    active: null
  }),
  methods: {
    onBack () {
      this.$store.dispatch('SET_TAB', null)
      this.$router.go(-1)
    },
    onNewTab () {
      this.$store.dispatch('SET_TAB', this.active)
    }
  }
}
</script>
