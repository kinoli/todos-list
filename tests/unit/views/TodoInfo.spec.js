import Vue from 'vue'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import TodoInfo from '@/views/TodoInfo.vue'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
Vue.use(Vuetify)

const localVue = createLocalVue()
localVue.use(Vuex)

const $router = {
  go: jest.fn()
}

const $route = {
  params: {
    id: '111'
  }
}

const mockState = {
  todos: [
    {
      userId: 2,
      id: 1,
      title: 'one two three',
      completed: false,
      user: {
        id: 2,
        name: 'Leanne Graham',
        username: 'policy',
        address: {
          street: 'Pennsilvania Avenue',
          city: 'Washington D.C.',
          zipcode: '00000'
        },
        company: {
          name: 'Foo'
        }
      }
    }
  ],
  tab: null
}

const mockTabClick = tab => {
  wrapper.vm.active = tab
  wrapper.vm.onNewTab()
}

let wrapper, mutations, actions, getters, store

const initWrapper = state => {
  actions = {
    SET_TAB: jest.fn()
  }
  getters = {
    getTodoDetails: () => () => (state.todos[0]),
    getActiveTab: () => (state.tab)
  }
  store = new Vuex.Store({
    state: state,
    mutations,
    actions,
    getters
  })
  wrapper = shallowMount(TodoInfo, {
    store,
    localVue,
    mocks: {
      $route,
      $router
    }
  })
}

describe('TodoInfo.vue', () => {
  it('renders correctly', () => {
    initWrapper(mockState)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('goes to previous page when back button is clicked', () => {
    initWrapper(mockState)
    wrapper.vm.$router = $router
    wrapper.find('#btn-back').vm.$emit('click')
    expect($router.go).toHaveBeenCalledWith(-1)
  })

  it('stores the clicked tab', () => {
    initWrapper(mockState)
    mockTabClick('User')
    expect(actions.SET_TAB).toHaveBeenCalled()
  })

  it('initializes with the correct preselected tab', () => {
    const state = {
      ...mockState,
      tab: 'User'
    }
    initWrapper(state)
    expect(wrapper.vm.active).toEqual('User')
  })
})
