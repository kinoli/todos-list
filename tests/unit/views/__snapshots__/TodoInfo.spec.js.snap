// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInfo.vue renders correctly 1`] = `
<section class="todo-info">
  <v-layout-stub tag="div" row="">
    <v-flex-stub tag="div" xs12="" sm6="" offset-sm3="">
      <v-card-stub>
        <v-toolbar-stub flat="true" scrollthreshold="300" tabs="true" dark="" color="teal">
          <v-btn-stub activeclass="v-btn--active" icon="true" tag="button" type="button" id="btn-back">
            <v-icon-stub>chevron_left</v-icon-stub>
          </v-btn-stub>
          <v-toolbar-title-stub></v-toolbar-title-stub>
          <v-spacer-stub></v-spacer-stub>
        </v-toolbar-stub>
        <!---->
      </v-card-stub>
    </v-flex-stub>
  </v-layout-stub>
</section>
`;
